File: javascript/linkedlist.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  createNode(value) {
    return new Node(value);
  }

  append(value) {
    let node = this.createNode(value);

    if (this.head === null) {
      this.head = node;
      this.tail = node;
      this.count = 1;
      return;
    }

    this.tail.next = node;
    this.tail = node;
    return;
  }

  insertAt(position, value) {
    if (this.head === null && position !== 1) {
      console.log("Invalid position");
      return;
    }

    if (position < 1) {
      console.log("Invalid posititon");
      return;
    }

    let node = this.createNode(value);

    if (this.head === null && position === 1) {
      this.head = node;
      this.tail = node;
      return;
    }

    let count = 0;
    let current = this.head;

    while (current.next !== null && count !== position - 2) {
      current = current.next;
      count = count + 1;
    }

    if (current.next !== null && count === position - 2) {
      node.next = current.next;
      current.next = node;
      return;
    }

    if (current.next === null && count === position - 2) {
      current.next = node;
      this.tail = node;
      return;
    }

    if (current.next === null && count !== position - 2) {
      console.log("Invalid position");
      return;
    }
  }

  prepend(value) {
    let node = this.createNode(value);
    if (this.head === null) {
      this.head = node;
      this.tail = node;
      return;
    }

    node.next = this.head;
    this.head = node;
    return;
  }

  remove(value) {
    if (this.head.value === value && this.tail.value === value) {
      this.head = null;
      this.tail = null;
      return;
    }

    if (this.head.value === value) {
      let temp = this.head;
      this.head = this.head.next;
      temp.next = null;
      return;
    }

    let current = this.head;

    while (current.next !== null && current.next.value !== value) {
      current = current.next;
    }

    if (current.next !== null && current.next.value === value) {
      if (current.next.next === null) {
        this.tail = current;
        current.next = null;
        return;
      }

      let temp = current.next;
      current.next = current.next.next;
      temp.next = null;
      return;
    }
  }

  removeAt(position) {
    if (this.head === null) {
      console.log("Invalid posititon");
      return;
    }

    if (position < 1) {
      console.log("Invalid posititon");
      return;
    }

    if (this.head.next === null && position !== 1) {
      console.log("Invalid position");
      return;
    }

    if (position === 1 && this.head.next === null) {
      this.head = null;
      this.tail = null;
      return;
    }

    if (position === 1 && this.head !== null) {
      let temp = this.head;
      this.head = this.head.next;
      temp.next = null;
      return;
    }

    let current = this.head;
    let count = 0;

    while (current.next !== null && count !== position - 2) {
      current = current.next;
      count = count + 1;
    }

    if (current.next === null) {
      console.log("Invalid position");
      return;
    }

    if (current.next !== null && count === position - 2) {
      if (current.next.next === null) {
        this.tail = current;
        current.next = null;
        return;
      }
      let temp = current.next;
      current.next = current.next.next;
      temp.next = null;
      return;
    }
  }

  pop() {
    if (this.head === null) {
      console.log("List is empty");
      return;
    }

    if (this.head.next === null) {
      this.head = null;
      this.tail = null;
      return;
    }

    let current = this.head;
    while (current.next.next !== null) {
      current = current.next;
    }

    this.tail = current;
    current.next = null;
    return;
  }

  shift() {
    if (this.head === null) {
      console.log("List is empty");
      return;
    }

    if (this.head.next === null) {
      this.head = null;
      this.tail = null;
      return;
    }

    let temp = this.head;
    this.head = this.head.next;
    temp.next = null;
    return;
  }

  getFirst() {
    if (this.head === null) {
      console.log("List is empty");
      return;
    }

    return this.head.value;
  }

  getLast() {
    if (this.head === null) {
      console.log("List is empty");
      return;
    }

    return this.tail.value;
  }

  isEmpty() {
    return this.head === null;
  }

  contain(value) {
    if (this.isEmpty()) {
      console.log("List is empty");
      return;
    }

    if (this.head.value === value || this.tail.value === value) {
      return true;
    }

    let current = this.head;

    while (current !== null) {
      if (current.value === value) {
        return true;
      }
      current = current.next;
    }

    return false;
  }

  get(index) {
    if (this.isEmpty()) {
      console.log("List is empty");
      return;
    }

    if (index < 0) {
      console.log("Invalid index");
      return;
    }

    if (this.head.next === null && index !== 0) {
      console.log("Invalid index");
      return;
    }

    if (this.head.next === null && index === 0) {
      return this.head.value;
    }

    let current = this.head;
    let count = 0;

    while (current !== null && count !== index) {
      current = current.next;
      count = count + 1;
    }

    if (current !== null && count === index) {
      return current.value;
    }

    if (current === null) {
      console.log("Invalid index");
      return;
    }
  }

  traverse() {
    let current = this.head;

    while (current !== null) {
      console.log(current.value);
      current = current.next;
    }
  }
}

module.exports = LinkedList;

File: src/linked-list.cpp
#include <iostream>
#include "node.cpp"
using namespace std;

template <typename T>
class LinkedList
{
private:
    Node<T> *head;
    Node<T> *tail;
    size_t length = 0;

public:
    LinkedList()
    {
        this->head = nullptr;
        this->tail = nullptr;
    }

    void push_back(T data)
    {
        Node<T> *node = new Node<T>(data);
        if (head == nullptr)
        {
            this->head = node;
        }
        else
        {
            this->tail->next = node;
        }
        this->tail = node;
        this->length++;
    }

    void pop_back()
    {
        if (this->head == nullptr)
        {
            throw out_of_range("Stack Underflow!");
        }

        if (this->head->next == nullptr)
        {
            free(this->head);
            this->head = nullptr;
            this->tail = nullptr;
        }
        else
        {
            Node<T> *temp = this->head;

            while (temp->next != this->tail)
            {
                temp = temp->next;
            }
            free(this->tail);
            temp->next = nullptr;
            this->tail = temp;
        }
        this->length--;
    }

    void push_front(T data)
    {
        Node<T> *node = new Node<T>(data);
        if (this->head == nullptr)
        {
            this->head = node;
            this->tail = node;
        }
        else
        {
            node->next = this->head;
            this->head = node;
        }
        this->length++;
    }

    void pop_front()
    {
        if (this->head == nullptr)
        {
            throw out_of_range("Stack Underflow!");
        }

        if (this->head->next == nullptr)
        {
            free(this->head);
            this->head = nullptr;
            this->tail = nullptr;
        }
        else
        {
            Node<T> *node = this->head;
            this->head = this->head->next;
            free(node);
        }
        this->length--;
    }

    void insert(size_t index, T data)
    {
        if (index > this->len())
        {
            throw out_of_range("Invalid index.");
        }

        if (index == 0)
        {
            push_front(data);
        }
        else if (index == this->len())
        {
            push_back(data);
        }
        else
        {
            Node<T> *temp = head;

            for (size_t i = 0; i < index - 1; i++)
            {
                temp = temp->next;
            }
            Node<T> *node = new Node<T>(data);
            node->next = temp->next;
            temp->next = node;
            this->length++;
        }
    }

    Node<T> *search(T data)
    {
        if (this->head == nullptr)
        {
            throw out_of_range("List is empty!");
        }
        else
        {
            Node<T> *temp = this->head;

            while (temp)
            {
                if (temp->data == data)
                {
                    return temp;
                }
                temp = temp->next;
            }

            return nullptr;
        }
    }

    size_t len()
    {
        return this->length;
    }

    void traverse()
    {
        Node<T> *temp = head;

        while (temp)
        {
            cout << temp->data << " -> ";
            temp = temp->next;
        }
        cout << "NULL" << endl;
    }

    void remove(T data)
    {
        if (this->head == nullptr)
        {
            throw out_of_range("List is empty!");
        }

        if (this->head->next == nullptr)
        {
            if (this->head->data == data)
            {
                pop_front();
            }
            else
            {
                throw out_of_range("Data not found!");
            }
        }
        else
        {
            Node<T> *temp = this->head;

            if (temp->data == data)
            {
                pop_front();
                return;
            }

            while (temp->next)
            {
                if (temp->next->data == data)
                {
                    Node<T> *node = temp->next;
                    temp->next = temp->next->next;
                    free(node);
                    this->length--;
                    return;
                }
                temp = temp->next;
            }

            throw out_of_range("Data not found!");
        }
    }

    void removeAt(int index)
    {
        if (index >= this->len())
        {
            throw out_of_range("Invalid index.");
        }

        if (index == 0)
        {
            pop_front();
        }
        else if (index == this->len() - 1)
        {
            pop_back();
        }
        else
        {
            Node<T> *temp = head;

            for (size_t i = 0; i < index - 1; i++)
            {
                temp = temp->next;
            }
            Node<T> *node = temp->next;
            temp->next = temp->next->next;
            free(node);
            this->length--;
        }
    }
};
